File: AndroidPromotionSDK/src/main/java/com/ttcn/promotionsdk/promotionsdkui/theme/PromotionThemeDefaults.kt
package com.ttcn.promotionsdk.promotionsdkui.theme

import com.ttcn.promotionsdk.promotionsdkui.theme.token.PRMButtonToken
import com.ttcn.promotionsdk.promotionsdkui.theme.token.PRMDiscountBadgeToken
import com.ttcn.promotionsdk.promotionsdkui.theme.token.PRMListItemToken
import com.ttcn.promotionsdk.promotionsdkui.theme.token.PRMSearchBarToken
import com.ttcn.promotionsdk.promotionsdkui.theme.token.PRMTabChipToken
import com.ttcn.promotionsdk.promotionsdkui.theme.token.PRMTabUnderlineToken

import android.content.Context
import androidx.annotation.ColorInt
import com.ttcn.promotionsdk.R
import com.ttcn.promotionsdk.promotionsdkui.utils.enum.PRMCoreButtonType
import com.ttcn.promotionsdk.promotionsdkui.widget.PRMSearchType
import com.ttcn.promotionsdk.promotionsdkui.utils.extension.retrieveColor

/**
 * Internal SDK defaults extracted from component source / resources.
 * Used by theme preview and when resolving unset token properties.
 */
internal object PromotionThemeDefaults {

    /**
     * Defaults aligned with [PRMButton] PRIMARY (common CTA in promotion screens).
     * Background is gradient in SDK; token stores gradient end color [R.color.tokenRainbowRedEnd].
     * Corner radius follows foundation button drawables ([R.dimen.tokenBorderRadius24] → _19sdp).
     * Note: sdp scales on device (e.g. ~25dp on a wide screen); that is the physical radius to apply via [PRMButtonToken].
     */
    fun button(context: Context) = PRMButtonToken(
        backgroundColor = context.retrieveColor(R.color.tokenRainbowRedEnd),
        textColor = context.retrieveColor(PRMCoreButtonType.PRIMARY.textColorRes),
        shadowColor = context.retrieveColor(R.color.tokenShadowsButtonColor),
        cornerRadius = pxToDp(context, context.resources.getDimension(R.dimen.tokenBorderRadius24))
    )

    fun searchBar(context: Context) = PRMSearchBarToken(
        borderColor = context.retrieveColor(R.color.tokenDark10),
        hintTextColor = context.retrieveColor(PRMSearchType.BASIC.hintColorRes),
        textColor = context.retrieveColor(PRMSearchType.BASIC.textColorRes),
        iconColor = context.retrieveColor(R.color.tokenDark40),
        cornerRadius = pxToDp(context, context.resources.getDimension(R.dimen.tokenBorderRadius08)),
    )

    fun listItem(context: Context) = PRMListItemToken(
        linkTextColor = context.retrieveColor(R.color.color_EE0033),
        usedBadgeTextColor = context.retrieveColor(R.color.tokenDark100),
        usedBadgeBackgroundColor = context.retrieveColor(R.color.color_f4f4f4),
        radioButtonStrokeColor = context.retrieveColor(R.color.tokenDark40),
        radioButtonSelectedStrokeColor = context.retrieveColor(R.color.color_EE0033),
    )

    fun tabChip(context: Context) = PRMTabChipToken(
        activeBackgroundColor = context.retrieveColor(R.color.color_4e4e4e),
        inactiveBackgroundColor = context.retrieveColor(R.color.color_f4f4f4),
        activeTextColor = context.retrieveColor(R.color.white),
        inactiveTextColor = context.retrieveColor(R.color.color_7A7A7A),
        cornerRadius = pxToDp(context, context.resources.getDimension(R.dimen.view_size_7)),
    )

    fun tabUnderline(context: Context) = PRMTabUnderlineToken(
        indicatorColor = context.retrieveColor(R.color.color_red_EE0033),
        activeTextColor = context.retrieveColor(R.color.black),
        inactiveTextColor = context.retrieveColor(R.color.color_7a7a7a),
        backgroundColor = context.retrieveColor(R.color.color_FBFBFB),
    )

    fun discountBadge(context: Context) = PRMDiscountBadgeToken(
        availableTextColor = context.retrieveColor(R.color.tokenPineBlue100),
        unavailableTextColor = context.retrieveColor(R.color.color_7A7A7A),
        availableBackgroundColor = context.retrieveColor(R.color.tokenPineBlue10),
        unavailableBackgroundColor = context.retrieveColor(R.color.color_f4f4f4),
        actionTextColor = context.retrieveColor(R.color.color_EE0033),
    )

    fun theme(context: Context) = PRMSDKTheme(
        buttonToken = button(context),
        searchBarToken = searchBar(context),
        listItemToken = listItem(context),
        tabChipToken = tabChip(context),
        tabUnderlineToken = tabUnderline(context),
        discountBadgeToken = discountBadge(context),
    )

    fun colorToHex(@ColorInt color: Int): String = ThemeHex.format(color)

    fun pxToDp(context: Context, px: Float): Float =
        px / context.resources.displayMetrics.density
}

File: iosPromotionSDK/PromotionSDKUI/Theme/PromotionThemeDefaults.swift
//
//  PromotionThemeDefaults.swift
//  PromotionSDK
//
//  Giá trị mặc định của SDK cho màn cấu hình theme của host: token nào chưa set thì hiện giá trị này.
//  Đối ứng `PromotionThemeDefaults.kt`.
//
//  **Màu khớp Android.** Dùng token PRMDesignKit có giá trị **bằng** `R.color.*` bên Android:
//
//    tokenViettelPayRed100 = #EE0033   tokenWhite   = #FFFFFF   tokenDark02 = #FBFBFB
//    tokenDark10  = #E9E9E9  tokenDark40 = #A7A7A7   tokenDark100 = #222222
//    tokenDark05  = #F4F4F4  tokenDark80 = #4E4E4E   tokenDark60  = #7A7A7A
//    tokenBlack   = #000000  tokenPineBlue100 = #2CA196   tokenPineBlue10 = #EAF6F4
//
//  Dùng token thay hex thô: nếu design system đổi token, default đi theo. `cornerRadius` giữ pt cố
//  định (Android dùng sdp co giãn theo màn, không có một giá trị duy nhất).
//

import UIKit
@_implementationOnly import PRMDesignKit

enum PromotionThemeDefaults {

    /// Nút chính. `backgroundColor` = màu cuối gradient; `cornerRadius` viên thuốc (`height/2` = 24 ở
    /// size lớn); `shadowColor` đỏ + opacity 25% = `#40EE0033` (khớp `tokenShadowsButtonColor`).
    static var button: ButtonToken {
        ButtonToken(
            backgroundColor: Colors.tokenViettelPayRed100,
            textColor: Colors.tokenWhite,
            shadowColor: Colors.tokenViettelPayRed100.withAlphaComponent(0.25),
            cornerRadius: 24
        )
    }

    /// Ô tìm kiếm. `iconColor` không có default thật (SDK render ảnh gốc); lấy `tokenDark40` cho khớp
    /// Android và để preview có gì đó hiển thị.
    static var searchBar: SearchBarToken {
        SearchBarToken(
            borderColor: Colors.tokenDark10,
            hintTextColor: Colors.tokenDark40,
            textColor: Colors.tokenDark100,
            iconColor: Colors.tokenDark40,
            cornerRadius: 8
        )
    }

    /// Item voucher. `linkTextColor` và hai màu radio bên iOS lấy từ ảnh asset; giá trị dưới đây khớp
    /// Android.
    static var listItem: ListItemToken {
        ListItemToken(
            linkTextColor: Colors.tokenViettelPayRed100,
            usedBadgeTextColor: Colors.tokenDark100,
            usedBadgeBackgroundColor: Colors.tokenDark05,
            radioButtonStrokeColor: Colors.tokenDark40,
            radioButtonSelectedStrokeColor: Colors.tokenViettelPayRed100
        )
    }

    /// Tab dạng chip.
    static var tabChip: TabChipToken {
        TabChipToken(
            activeBackgroundColor: Colors.tokenDark80,
            inactiveBackgroundColor: Colors.tokenDark05,
            activeTextColor: Colors.tokenWhite,
            inactiveTextColor: Colors.tokenDark60,
            cornerRadius: 7
        )
    }

    /// Tab gạch chân.
    static var tabUnderline: TabUnderlineToken {
        TabUnderlineToken(
            indicatorColor: Colors.tokenViettelPayRed100,
            activeTextColor: Colors.tokenBlack,
            inactiveTextColor: Colors.tokenDark60,
            backgroundColor: Colors.tokenDark02
        )
    }

    /// Badge giảm giá.
    static var discountBadge: DiscountBadgeToken {
        DiscountBadgeToken(
            availableTextColor: Colors.tokenPineBlue100,
            unavailableTextColor: Colors.tokenDark60,
            availableBackgroundColor: Colors.tokenPineBlue10,
            unavailableBackgroundColor: Colors.tokenDark05,
            actionTextColor: Colors.tokenViettelPayRed100
        )
    }

    /// Theme mặc định đầy đủ. Đối ứng `PromotionThemeDefaults.theme(context)` bên Android.
    static var theme: PromotionSDKTheme {
        PromotionSDKTheme(
            buttonToken: button,
            searchBarToken: searchBar,
            listItemToken: listItem,
            tabChipToken: tabChip,
            tabUnderlineToken: tabUnderline,
            discountBadgeToken: discountBadge
        )
    }

    /// `#AARRGGBB`, rút gọn `#RRGGBB` khi màu đục. Đối ứng `PromotionThemeDefaults.colorToHex`.
    static func colorToHex(_ color: UIColor) -> String { ThemeHex.format(color) ?? "#000000" }
}
